src/commands/test/runner.py

```diff
@@ -150,22 +166,44 @@ async def invoke_test_function(self, function_name: str):
         )
 
         func = getattr(self.test_contract, function_name)
+        is_failure_expected = (
+            function_name.startswith("test_fail_") and self._is_test_fail_enabled
+        )
+
         # TODO: Improve stacktrace
         try:
-            call_result = await func().invoke()
-            if self._is_test_error_expected:
-                raise MissingExceptReportedException(
-                    "Expected a transaction to be reverted"
+            try:
+                call_result = await func().invoke()
+                if self._expected_error is not None:
+                    raise MissingExceptReportedException(
```

Comment:
I would advise strongly to thoroughly test this function if we're not going to be simplifying the logic, so when other people try to tweak the logic we'll have a clear idea what the cases are here.